_posts/2020-09-22-rust-2021.md

```diff
@@ -0,0 +1,96 @@
+
+I love research more than most anything, and a lot of my own work has a strong research flavor. That has caused some confusion; some of the things I'm exploring are very speculative and will likely take years to come to fruition. Certainly my research into compute-centric GPU rendering is of that nature; I'm excited about the fact that it promises dramatic performance improvements over the current state of the art, but it's nowhere near ready to put into production yet. I'm striving for clearer communication so people can have a better idea what is speculative, based on grand futuristic visions, and what is on track to being usable reasonably soon. But both are important aspects of what I consider to be the main mission: fostering learning about how to build GUI.
+
+## Baby steps
+
+While I am driven by a long-term, ambitious vision, the goal of Druid in 2021 is not to deliver a general UI toolkit. Rather, we are deliberately continuing to follow a narrow scope. The primary goal remains the font editor project, and we plan to re-focus attention on that. I do think this is an attainable goal. I also think that what we learn from trying to build a real application with users will be extremely valuable to the more ambitious task.
+
+One project management technique that is proving effective is "cycles." Instead of trying to solve the most ambitious version of a problem, we choose up front what to push to a future cycle, reducing the scope for the current implementation cycle. An example is the choice to defer BiDi from our recent text work. This is obviously an essential feature for a real GUI toolkit, but we also know it could take weeks or months to get it right. To have any chance of shipping, we have to carefully budget our time and energy on subprojects that easily could expand to absorb our full attention.
+
+## A wishlist
+
+Here are some things I'd especially like from the rest of the Rust community.
+
+First, I'd love for wgpu to become more mature, as I think it is the most promising approach to GPU for the future. Our [roadmap for 3D] basically depends on it. Much love to that community for the progress they've made so far and hope for the future.
+
+Second, while we generally find Windows platform bindings to be straightforward, thanks to the excellent winapi bindings (and hopefully a stable [com-rs] shortly), the situation on macOS is not as good. There are real problems interoperating with Objective-C code, including type-safety in wrappers that also allow subclassing, and continual friction around autorelease pools. There are also different, incompatible conventions around wrapping foreign types (foreign_object vs TCFType). None of this is blocking, but it makes developing macOS platform integration less fun. (See [this HN thread](https://news.ycombinator.com/item?id=24309565) for a bit more background)
```

Comment:
Agreed. Dealing with macOS bindings has been quite a painful experience for me too when working on [`winit`](https://github.com/rust-windowing/winit).

This situation is also one of the main reasons I stopped working on [`piet-window`](https://github.com/hecrj/piet-window).